src/order/partition/finpartition.lean

```diff
@@ -0,0 +1,486 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import algebra.big_operators.basic
+import order.locally_finite
+import order.atoms
+import order.sup_indep
+
+/-!
+# Finite partitions
+
+In this file, we define finite partitions. A finpartition of `a : α` is a finite set of pairwise
+disjoint parts `parts : finset α` which does not contain `⊥` and whose supremum is `a`.
+
+## Constructions
+
+We provide many ways to build finpartitions:
+* `finpartition.of_erase`: Builds a finpartition by erasing `⊥` for you.
+* `finpartition.of_subset`: Builds a finpartition from a subset of the parts of a previous
+  finpartition.
+* `finpartition.empty`: The empty finpartition of `⊥`.
+* `finpartition.indiscrete`: The indiscrete, aka trivial, aka pure, finpartition made of a single
+  part.
+* `finpartition.discrete`: The discrete finpartition of `s : finset α` made of singletons.
+* `finpartition.bind`: Puts together the finpartitions of the parts of a finpartition into a new
+  finpartition.
+* `finpartition.atomise`: Makes a finpartition of `s : finset α` by breaking `s` along all finsets
+  in `F : finset (finset α)`. Two elements of `s` belong to the same part iff they belong to the
+  same elements of `F`.
+
+`finpartition.indiscrete` and `finpartition.bind` together form the monadic structure of
+`finpartition`.
+
+## Implementation notes
+
+Forbidding `⊥` as a part follows mathematical tradition and is a pragmatic choice concerning
+operations on `finpartition`. Not caring about `⊥` being a part or not breaks extensionality (it's
+not because the parts of `P` and the parts of `Q` have the same elements that `P = Q`). Enforcing
+`⊥` to be a part makes `finpartition.bind` uglier and doesn't rid us of the need of
+`finpartition.of_erase`.
+
+## TODO
+
+Link `finpartition` and `setoid.is_partition`.
+
+The order is the wrong way around to make `finpartition a` a graded order. Is it bad to depart from
+the literature and turn the order around?
+-/
+
+open finset function
+open_locale big_operators
+
+variables {α : Type*}
+
+/-- A finite partition of `a : α` is a pairwise disjoint finite set of elements whose supremum is
+`a`. We forbid `⊥` as a part. -/
+@[ext, derive decidable_eq] structure finpartition [lattice α] [order_bot α] (a : α) :=
+(parts : finset α)
+(sup_indep : parts.sup_indep id)
+(sup_parts : parts.sup id = a)
```

Comment:
How often does it happen that you want to delete or add some elements from/to a partition? In future work, will you want to move between `finpartition`s of `α` and `option α` or things like that?
If this deleting/adding is something that is a useful feature in combinatorics (which wouldn't surprise me), then I wonder if it's better to add one more layer here. Define a `finpartion (s : finset α)` of arbitrary finite sets of elements of a random (possibly infinite) ambient type `α`.

Reply:
This definition already does this! `a` is the element which is partitioned, and it is not required to be `⊤`. In practice, and that's what you're thinking about, we'll substitute `finset α` for `α`.

Reply:
Aah, that shows how well I understood what's going on here...